### PR TITLE
Manage search synonyms

### PR DESCRIPTION
**Pull request by aboredman**
| Q               | A
| --------------- | ---
| Bug fix?        | no 
| New feature?    |  yes
| New sample?      | no
| Related issues?  | first implementation of #100 

#### What's in this Pull Request?

This PR adds a new feature allowing users to manage synonyms used in a search query.
The synonyms are defined in a properties of the SearchResults WP, are loaded onInit in a synonym table and are added to the querytext right before it is sent to the server.

This version is heavily based on the work by @wobba and @estruyf

The implementation is working fine but many improvements could be added (use RefinementFilters instead of the QueryText to avoid length limits, allow synonyms import from a JSON or similar file, refactor of some elements of code to increase performance, etc.). I plan to keep working on these features but I think that the current version could be of use.

Thanks for your contribution! Sharing is caring.

### CODE REVIEW

**Comment by FranckyC** [comment]
Hi @aboredman, seems working as expected ;). Thank you for the update. If you could make few minor updates before merging, would ne nice :
- Remove unused localization labels ('SynonymJSON')
- Change property to camel case in ISynonymFieldConfiguration.ts
- Please use an interface for synonymTable instead of `{ [key:string]: string[] }`
- As suggested by @wobba, please add synonyms as FQL refinement filter (like `or(term1,term2)` etc.) instead of KQL. It will be easier to manage.
- Replace 'List of synonyms' label by 'Configure synonyms' in the property pane property
- In `PropertyFieldCollectionData('synonymList', {...`, use localized labels

**Comment by FranckyC** [comment]
I also forgot:
- Update the README file

File: solutions/ModernSearch/react-search-refiners/spfx/src/services/SearchService/SearchService.ts [hunk 1]
```diff
@@ -383,6 +388,60 @@ class SearchService implements ISearchService {
...
     }
+
+    // Function to inject synonyms at run-time
+    private _injectSynonyms(query: string): string {
+        // Remove complex query parts AND/OR/NOT/ANY/ALL/parenthasis/property queries/exclusions - can probably be improved            
+        const cleanQuery = query.replace(/(-\w+)|(-"\w+.*?")|(-?\w+[:=<>]+\w+)|(-?\w+[:=<>]+".*?")|((\w+)?\(.*?\))|(AND)|(OR)|(NOT)/g, '');
+        const queryParts: string[] = cleanQuery.match(/("[^"]+"|[^"\s]+)/g);
+
+        // code which should modify the current query based on context for each new query
+        if (queryParts && this._synonymTable) {
```
Comment (wobba) [inline solutions/ModernSearch/react-search-refiners/spfx/src/services/SearchService/SearchService.ts]: Hi @aboredman , can you move the table check to line 395 and check for null, and size of 0. Saves doing regex on the query if there are no synonyms. Minor, but better 🙂

File: solutions/ModernSearch/react-search-refiners/spfx/src/webparts/searchResults/SearchResultsWebPart.ts [hunk 2]
```diff
@@ -263,6 +266,34 @@ export default class SearchResultsWebPart extends BaseClientSideWebPart<ISearchR
         });
     }
 
+    private _convertToSynonymTable(synonymList: ISynonymFieldConfiguration[]): { [key: string]: string[] } {
+        let synonymsTable: { [key: string]: string[] } = {};
```
Comment (wobba) [inline solutions/ModernSearch/react-search-refiners/spfx/src/webparts/searchResults/SearchResultsWebPart.ts]: Do we need a null check here if upgrading and synonymlist is not there?

**Comment by wobba** [comment]
I added two minor code comments. I'll merge this without doing FQL as a first implementaion if you find it hard to do, as you copied my KQL expansion code 😉

**Pull request merged**